wallet rescan: Add error for SPV mode and excessive depth in pruned mode

pinheadmz opened the pull request:
In SPV mode, `rescan` doesn't actually do anything. If a user creates a watch-only wallet by importing an `xpub` while in SPV mode, that wallet will not necessarily be in sync with the chain. Unlike fullnode, a `rescan` will not sync the wallet. This PR adds an error that redirects the user to `reset` which, in SPV mode at least, should not take too long and will discover any wallet transactions that occurred before the import point.

This issue also applies to pruned nodes, if historical transactions for the imported address were confirmed before the prune height. However a pruned node can not `reset`. We could expand the error to cover pruned nodes as well.

We might also want to add a note in the docs that importing keys into a pruned node may not correctly recover those keys' blockchain history.

Or... is it possible to enable `reset` on a pruned node?

pinheadmz commented:
Actually this may be a watch-only wallet issue, not an SPV issue. Having the same behavior importing an `xpub` on a full node as well.

tuxcanfly commented:
Needs a rebase to fix tests.

pinheadmz commented:
Rebased. Tests successful on local. Think I will want to add one more commit to check prune height against `rescan` requested height, but would need to import `chain` into `wallet/rpc.js` (See https://github.com/bcoin-org/bcoin/issues/536)

tuxcanfly commented:
I saw `reset` was getting timed out, presumably because the operation takes longer than bclients timeout.

pinheadmz commented:
I think I noticed that too. The command times out but the rest continues until it's done in the background.

pinheadmz commented:
Adding to this conversation: There should be an error if a pruned node is asked to rescan beyond it's prune depth. I have a tx in my wallet that is stored as `unconfirmed` even though it was confirmed 1,000 blocks ago. 

What happened, I think, is that I executed `rescan` with a height deeper than my prune depth. The wallet clears all confirmations shallower than the given height (without any checks), but on a pruned node, the block confirming my tx was no longer on disk, so it did not get reconfirmed during the `rescan`.

pinheadmz commented:
UPDATE: This PR now adds an important error if wallet rescan is attempted on a pruned node with a depth greater than the number of blocks on disk.

A new test file was added `prune-test.js` -- I'm sure we have more tests we can do on pruned nodes. I would also add an `spv-test.js` file, but it's pretty tricky generating blocks on a SPV node!

This PR also updates the regtest block parameters around pruning to make testing more practical:

```diff
--- a/lib/protocol/networks.js
+++ b/lib/protocol/networks.js
@@ -775,8 +775,8 @@ regtest.block = {
   bip65hash: null,
   bip66height: 1251,
   bip66hash: null,
-  pruneAfterHeight: 1000,
-  keepBlocks: 10000,
+  pruneAfterHeight: 500,
+  keepBlocks: 288,
   maxTipAge: 0xffffffff,
   slowHeight: 0
 };
```

pinheadmz commented:
rebased on master, testing against `bmocha`

pinheadmz commented:
@nodar-chkuaselidze @braydonf Did some updates:

- bcoin node `getInfo()` now also returns `chain` properties `spv` and `prune` (booleans).

- All wallet node clients (`nodeclient`, `nullclient` and `client` in lib/wallet) have new method `getInfo()` that returns the bcoin client getInfo chain properties

- All wallet types (plugin, independent node, and walletDB with no node & `nullclient`) check `getinfo` during `open()` and store the flags in a new object `this.chain`

tuxcanfly commented on lib/wallet/http.js:
```diff
@@ -190,9 +191,23 @@ class HTTP extends Server {
+        const minHeight = tipHeight.height - pruneDepth;
+
+        enforce(height > minHeight,
+          'Cannot rescan past prune depth of ' + minHeight);
```
Should we say "rescan height must be greater than minHeight"? It's a bit clearer and more direct. Bit confusing to have to think about depth when the param is height.

pinheadmz replied:
👍

tuxcanfly commented on lib/wallet/node.js:
```diff
@@ -108,6 +108,12 @@ class WalletNode extends Node {
     await this.handleOpen();
 
+    const clientInfo = await this.client.getInfo();
+    this.chain = {
```
I think we should a different var than `this.chain` for this.

pinheadmz replied:
Making this change in `walletdb.js` and `plugin.js`

tuxcanfly commented on test/prune-test.js:
```diff
@@ -0,0 +1,113 @@
+
+  it('should generate 1000 blocks', async () => {
+    const addr = await wallet.createAddress('default');
+    assert(await nclient.execute('generatetoaddress', [100, addr.address]));
```
Why not just 1000 at once?

pinheadmz replied:
Well - I was getting timeout errors. Not from the testing suite (because timeout is st to `60000` on line 37) but from actual bclient. And not even on local, just on Travis. SO it's a bit of a hacky work-around. I'm using bclient to generate blocks in this test, whereas other blockchain tests in `bcoin` use custom functions like `mineBlock` and `CPUminer`. I was lazy in writing this test with bclient but also figure, we're getting a tiny bit of extra coverage this way because there aren't any other tests for this API call.

tuxcanfly commented on lib/protocol/networks.js:
```diff
@@ -775,8 +775,8 @@ regtest.block = {
   bip66hash: null,
-  pruneAfterHeight: 1000,
-  keepBlocks: 10000,
+  pruneAfterHeight: 500,
```
Minor backward incompat note.

pinheadmz replied:
Oh yeah, add a note to `/CHANGELOG.md` ?

pinheadmz commented on lib/node/http.js:
```diff
@@ -125,7 +125,9 @@ class HTTP extends Server {
           tip: this.chain.tip.rhash(),
-          progress: this.chain.getProgress()
+          progress: this.chain.getProgress(),
+          spv: this.chain.options.spv,
```
I don't think those properties exist in a `chain` object (could be wrong) -- or are you saying they _should_ be, as part of this PR? They are properties of a `ChainOptions` object, also defined in `lib/blockchain/chain.js`

pinheadmz replied:
👍

pinheadmz replied:
OK actually there is no property `chain.spv` or `chain.prune`, those props are kept in `options` and never set at the "root" level of the object.

When I remove `.options.` the results are undefined and so they don't get returned by the `getInfo` request at all, and the test will always think prune and SPV are false.

Not sure how to test for this...?

pinheadmz replied:
Ok added test to check `prune` value from `getInfo()`

pinheadmz commented:
Tests are failing on timeouts generating blocks again, hm seems like sometimes it does sometimes it doesn't...

pinheadmz commented:
NICE! Little rebase-on-master fixed the testing issue

pinheadmz commented:
@tuxcanfly @braydonf 

Ok this PR got a little long, but I think it's in great shape now. Going to put a recap in this comment:

Motivation: Wallet can not rescan an SPV chain at all. Pruned nodes can only rescan the limited number of blocks on disk. In the case of a pruned node, rescanning beyond the depth of the chain stored on disk means transactions will be marked as "unconfirmed" forever.

Solution: Wallet asks chain what its properties are and throws an error in either case described.

Method:

1. There are three ways wallet connects to chain: `nullclient` (testing), `nodeclient` (plugin), and `client` (standalone server). Each of these three objects were updated to get `chain` properties from whatever chain is about to be rescanned.

2. To support this as well as inform the user, new properties `spv` and `prune` were added to the `chain` object in the http `getInfo()` call.

3. Testing. Created a new test module `prune-test.js` that tests a too-deep rescan command. I also changed the regtest params for pruning, and added this note in the draft CHANGELOG:

> In the old configuration, a pruned node on Regtest would start pruning at height
> 1000 and then keep the last 1000 blocks on disk (pruning anything older). This
> has been changed to facilitate testing. The new parameters are
> `pruneAfterHeight: 500, keepBlocks: 288`.

pinheadmz commented:
Closing this for https://github.com/bcoin-org/bcoin/pull/749 and I'll make another PR to actually handle the rescans.
